### ISSUE TITLE
Single source of truth for dependency lists.

### ISSUE THREAD

**Issue by stuaxo** [shoebot/shoebot#331]
It would be good to have one source of truth for the lists of packages, I'm interested to hear ideas on what this should be.

Right now, it's the `./install/install_dependencies.sh` script.
We have a unittest [1] that checks the package list for debian is the same as in our travis script.

Ultimately something else should be the source of truth.  The script isn't optimal if we want to do things in the future like show the user a hint `pip install shoebot` fails, and probably other use cases.

[1] Test that verifies the travis package list is in sync with the script:
https://github.com/shoebot/shoebot/blob/master/tests/unittests/test_travis_configuration.py#L48-L59

**Comment by stuaxo** [comment]
What are the minimum dependencies ?

Work for readthedocs makes me think it's a bit smaller than those in `install/install_dependencies.sh`.

**Comment by rlafuente** [comment]
Was thinking this a couple of days ago -- is GTK needed for the command line runner? 

If not, we could have a separate deps list if you don't need the gui, since some of those seem to be tricky to deal with

On February 24, 2021 12:45:15 AM GMT+00:00, Stuart Axon <notifications@github.com> wrote:
>What are the minimum dependencies ?
>
>Work for readthedocs makes me think it's a bit smaller than those in
>`install/install_dependencies.sh`.
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/shoebot/shoebot/issues/331#issuecomment-784642923

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by stuaxo** [comment]
It would be good to try the dependencies list that rtd uses and see what works (in theory everything headless should).

I'd like to adjust the tests, so that if the GUI stuff isn't available it automatically disables the GUI tests.

On a similar tip, I'm thinking of vendorizing the one function in [meta](https://pypi.org/project/meta/) that we use because it has an unapplied PR for python 3.8, and isn't available in debian.

Doing both of these things should make it a lot easier to make a distro package, something @speters33w was keen on for MINGW64.

**Comment by stuaxo** [comment]
TBH my way of specifying package lists has not been super scientific - add things until stuff works.

Is best practice to just have a list of the top level packages needed?

**Comment by rlafuente** [comment]
IIRC readthedocs was choking on not having `gobject-introspection`, which made me wonder at the time whether the GTK dependencies should be mandatory. I'm not sure whether it's worth the time to check with them if they can include GTK libs, which are *heavy*.

For headless operation (eg automated generation in servers, an use case i'm personally keen on) maybe we could have a minimal set of dependencies/requirements? I can try putting together a requirements-minimal.txt and see if that could work with RTD. Or just let me know if I should focus on something else.

**Comment by stuaxo** [comment]
In this version of shoebot we're always going to need a bunch of "G" libraries.  Gtk itself is not needed, but we need Glib and `gobject-introspection` is how we talk to some other libraries.


Issues installing using only the system package manager (lets say `apt-get` for now):

`meta` and [pubsub](https://pypi.org/project/pubsub/) are not available on debian, we can probably remove meta, but `pubsub` would be difficult.

`pubsub` is how we send messages through shoebot to do things like load new code or quit so would be hard to remove.

Packaging this for debian as `python3-pubsub` would be help: https://github.com/nehz/pubsub/issues/1 but I'd rather fix code issues than take the time to get this through the process.


Issues installing from pip:

`pycairo` and `pygobject` are not available as WHLs on Linux yet so have to build from source (and so need a bunch more dependencies).

cairo (via pycairo) brings in a lot more dependencies than I imagined:
- cairo is built as one library, so links with enabled for... since it's built with Gtk support you get Gtk and then the same applies - so you end up with X, fontconfig etc.

It's a a disaster if you want to minimise dependencies (e.g. if you are building in docker).

Under docker, some people build versions of cairo without Gtk support as a solution.

This would need a change in cairo itself to be fixed - e.g. split out the backends into plugins it loads at runtime, I don't have the C experience or the time to do this though.



So yeah - it's a nightmare, which is why I want to do something different after 1.4, but for this version we can't fix all the mess only some of it.
It looks like getting pubsub would be a win.
On the pycairo side, there is progress towards WHLs and that help from the `pip` installation side of things.

**Comment by rlafuente** [comment]
argh, i think it's probably the 4th time you explain this to me, now i remember there's no easy fix...

**Comment by stuaxo** [comment]
I had to do some digging to get the details - I really only learned the state of cairo this year.

I was unable to use cairo at my last work contract, there were too many dependencies baked in for our limited docker environment.

I ended up on the cairo mailinglist where I found out more details, if I had infinite time and motivation I'd try and fix it, but I don't :man_shrugging: 

Weasyprint: one of the big apps python that use Pango and Cairo is moving off the stack for similar reasons.   Part of me feels like this is a massive failure on Cairo as a cross platform library, but this is looking at it in the wrong way:

These libraries were built for Linux and X, and other platforms were added.
Cross platform as such wasn't a core aim (at least: no company stepped forward to pay someone to work on cross-platform, just to work on adding one-more platform).


Virtualenv, and later WHLs came along later, people building for them have different aims: to be self contained, to not need root for installation.


We should at least *try* to untangle some of dependency woes for this release.
I haven't got any experience of making debs, but I guess that is one thing we could try, another is Flatpak or AppImage (I guess we pack our IDE, not the Gedit plugin there).


After that we can start on a 2.x - one of the most important goals for me is being radically cross platform, with this limitation we could go to places that only specialist python libraries now - maybe we could even work in web based pythons such as Brython for instance.

Shoebot has really taught me, that software only exists as much as users can install it.


Actions:
My immediate action will be to look at removing the `meta` dependency (we only use 1 function from it).

For pubsub, we can either:
- [1] Make pubsub.deb and try and get this accepted into debian:
  This is my preference, but I'll probably need help.

- [2] Vendorize (copy it into shoebot)
  Straightforward in the short term, but would probably stop shoebot being accepted in debian until we did [1]

- [3] Remove the dependency.
  I'm not super keen on doing this sort of surgery on shoebot 1.x for something that isn't fixing a bug.

This would give us a route to shoebot 1.x as a deb, which solves one side of the dependency puzzle.


Apologies for the long text!

**Comment by rlafuente** [comment]
ok just a note FWIW, on debian testing i only had to install these for a successful install:

```
libpango1.0-dev (also pulls cairo)
libgirepository1.0-dev (also pulls gobject-introspection)
```

**Comment by stuaxo** [comment]
You are probably missing SVG support

**Comment by stuaxo** [comment]
I'd check what diagnose says to be sure.

**Comment by stuaxo** [comment]
There must be a tool that given a list of packages gives the definitive list to use as your dependencies ?

**Comment by rlafuente** [comment]
yep you're right, no svg support. All other boxes checked though

(but i tested this on my office system, which already has a lot of stuff installed, so this is not a good sample...)

**Comment by stuaxo** [comment]
Starting from a headless docker image is something I've done in the past.

**Comment by stuaxo** [comment]
OK, re: pubsub.   I'm starting an experimental branch to use pypubsub, it seems to work (all tests passing).

pypubsub is packaged for debian/ubuntu.. so should exist in all the other places.

**Comment by stuaxo** [comment]
OK, pypubsub swapped in - hopefully nothing broke.
